### ISSUE TITLE
kernel: add a LOCKED_TNUM object flag

### ISSUE THREAD

**Issue by fingolfin** [gap-system/gap#4143]
Right now, all non-immediate GAP objects can potentially change their TNUM. This can lead to subtle bugs, and also prevents certain optimizations. Having locked types would also allow improvements to the GAP-Julia bridge which are currently not possible (or at least risky -- I can of course already now assume that e.g. a `T_FUNCTION` never changes its TNUM, but right now somebody could still write code to thwart this.

I propose that we introduce a way to allocate bags which have a "locked" TNUM that never changes, and/or a function to "lock" the tnum of an existing object. This would be enforced by the `RetypeBag` or `PrecheckRetypeBag` C kernel function. 

First off, I would be to add a table of TNUMs which are *always* locked. This is akin to `MakeBagTypePublic` in HPC-GAP. Call this e.g. `MakeBagTypeLocked`.

This would already cover my personal uses case. But if desired, we could also allow individual bags to have this trait, by introducing a `OBJ_FLAG_LOCKED_TNUM` object flag. This could be useful for e.g. immutable strings (such as function names, etc.) kept in the kernel, which we currently cannot directly expose to user level code as that might end up calling `PLAIN_LIST` on them leading to havoc down the road. These strings could be type locked, then it'd be safe to expose them to the user.

Locking the type of an individual object could still can be done with just `SET_OBJ_FLAG(obj, OBJ_FLAG_LOCKED_TNUM)`. There would also be code in `CLEAR_BAG_FLAG` which enforces that this flag is never cleared (not much code is using `CLEAR_BAG_FLAG` resp. `CLEAR_OBJ_FLAG`, so I am not concerned about performance issues here).

Finally, there could be a `NewBagWithFlags` call; that would then also nicely fit with the possibility of creating immutable bags (once `IMMUTABLE` becomes an object flag, see issue #2325). However, I am not convinced that we really need this? In any case, it's easy enough to add later on.

The following TNUMs should probably always be locked:
- `T_INTPOS`
- `T_INTNEG`
- `T_RAT`
- `T_CYC`
- `T_MACFLOAT`
- `T_BOOL`
- `T_CHAR`
- `T_FUNCTION`
- `T_BODY`
- `T_FLAGS`
- `T_HVARS`
- `T_COMOBJ`
- `T_POSOBJ`
- `T_DATOBJ`
- `T_WPOBJ`

It's a bit less clear for (partial) permutations and transformations: `TRIM_PERM` etc. do retype them after all. But I guess at least the 2-byte versions could also be locked?

I guess a bunch of HPC-GAP specific ones like `T_ACOMOBJ`, `T_APOSOBJ`, `T_THREAD`, etc. could also be type locked, but I didn't think much about them.

Note that `T_LVARS` can change to `T_HVARS` so we can't have that.

**Comment by fingolfin** [comment]
Ping @ChrisJefferson @ThomasBreuer @rfourquet

**Comment by ChrisJefferson** [comment]
I'm happy with this. I would personally be happy with avoiding TRIM_PERM changing TNUM (we could make long term plans for this). it could just not change the type while triming (I don't think they would be a problem, just waste a little memory on the rare occasions trim pushes you from PERM4 to PERM2?)

**Comment by fingolfin** [comment]
In PR #4142 some concern about uses of `CLONE_OBJ`, `SWITCH_OBJ` were raised by @ChrisJefferson, and I analyzed this a bit; in the end, I think this is more related to locked TNUMs than to "constant" objects, so I am posting my findings here, not there.

I looked at were those functions are actually used, and found this
- to the best of my knowledge, no package uses these (at least not directly)
- in the kernel, `src/objset.c` uses them to implement `ClearObjSet`, `ResizeObjSet`, `ClearObjMap`, `ResizeObjMap`; but these are "safe" as far as I am concerned, as they don't change the TNUM.
- in the library, `InstallValue`, `InstallFlushableValue`, `InstallFlushableValueFromFunction` are the primary, critical users; this already has been causing us pain, see issue #1637 (and also #274)
  - once GAP 4.12 is out, we should work on getting as many packages as possible to use `DeclareGlobalName` instead of `DeclareGlobalVariable` resp. `DeclareGlobalFunction`; that won't solve the issue but reduce what it covers
  - currently the placeholder objects are `T_POSOBJ` and above I suggest to turn that into a locked TNUM: so there's a problem here. This could be resolved by adding a dedicated `T_PLACEHOLDER` TNUM (ugly as that would be). Or by not locking `T_POSOBJ` by default (that would be only a minor loss from my POV, overall)
  - for the "flushable" variant, I observe that all calls to it in the library and in packages install objects which are plain lists or records... and the overall type stays the same (although the TNUM might not, for lists, for the usual silly reason that we have multiple TNUMs for plists depending on their content). Anyway: if we turned this into a requirement, one could rewrite `InstallFlushableValue` and `InstallFlushableValueFromFunction` to not use `CLONE_OBJ` or `SWITCH_OBJ` or `FORCE_SWITCH_OBJ` (at least if one uses them in conjunction with `DeclareGlobalName` instead of `DeclareGlobalValue`, so that the initial value is also a plain list or record, not a placeholder object)
- `hpcgap/lib/hpc/queue.g` uses `SWITCH_OBJ` in `ExpandQueue`; but that's for mutable plists and so "fine" at least for my purposes
- `lib/vecmat.gi`, `hpcgap/lib/vecmat.gi` and `hpcgap/lib/vec8bit.gi` use them quite a bit. Not sure what to do there. This code has been another thorn in our side for some time. *sigh* How I wish MatrixObj had been done and finished and used 20 years ago, instead of this whole "automagic conversion of objects to "better" representations" stuff *doublesigh*. The critical part here is when this is used on a vector which is already compressed (so an external object) to change it to a different base field. 
- for HPC-GAP, `lib/type1.g` uses `FORCE_SWITCH_OBJ`. Not great. But this only happens if the inputs are (mutable) plain list or records, and so again this is fine from my POV.

All in all, the above suggest to me that for the locked TNUM PR, at first NOT locking the TNUMs for external objects would be more realistic.

**Comment by ChrisJefferson** [comment]
I would be in favour (although don't want to do the work...) of adding a T_PLACEHOLDER TNUM, because as you say this might be a thing we almost never get rid of, so making it a special TNUM we can special case in future seems sensible -- we might be able to say that is the only TNUM which can change eventually (ignoring things rotating around the different LIST tnums, and that type of thing)

**Comment by ThomasBreuer** [comment]
@fingolfin If I understand the sketch about the long-term perspectives right then you want to eventually get rid of `INSTALL_GLOBAL_FUNCTION`.
The feature to replace a function (to which many pointers may exist) by another one is used by the GAP library function `Debug`, and I do not know how to achieve this functionality without `INSTALL_GLOBAL_FUNCTION`.
(`Debug` is available only after calling `ReadLib( "debug.g" )`.)

**Comment by fingolfin** [comment]
No, `INSTALL_GLOBAL_FUNCTION` actually is fine, it doesn't modify the TNUM.

**Comment by ThomasBreuer** [comment]
What is then the problem with `DeclareGlobalFunction`? It returns a function, and `InstallGlobalFunction` replaces it by another function, with the same TNUM.

**Comment by fingolfin** [comment]
Never answered the final question asked here two years ago, so: There is no immediate problem with `DeclareGlobalFunction`, from my point of view it'd be OK if we keep using it.

**Comment by fingolfin** [comment]
I've come upon this again today, in the context of the Julia integration of GAP, where it would allow me to perform all kinds of cool things and optimizations if it weren't for the fact that GAP objects can magically change their type: that means I can never directly link GAP objects to Julia counterparts. In a dream world, I could do funky things like substituting Julia permutations for our classical GAP permutations (of course this would have other problems as code in the kernel and in various packages directly accesses `T_PERM2/4` objects, but these could be dealt with, if one is sufficiently motivated. Anyway, I have no immediate plans for *this*, just wanted to give an example).

Back to the original task: So far I only discussed "locking" certain types. But this doesn't work well for e.g. `T_PLIST*` because those can freely change between the various "subtypes". Which is actually *fine* for my purposes, because really they all have the same memory layout; we just encode some extra information in the TNUM. So then perhaps the right way to model this is to create a 256 x 256 type transition matrix which records for each TNUM what other TNUMs it may be retyped to "safely". I guess then `T_PLIST -> T_POSOBJ` and `T_REC -> T_COMOBJ` could also be allowed.

That would be relatively flexible and also allow us to move to a stricter regime gradually. It would also be orthogonal to more strict locking via an `OBJ_FLAG_LOCKED_TNUM` flag.
